File: playwright-python-ui/pages/sauce_demo_pages/login_page.py
from playwright.sync_api import Page, expect
from model.user import User


class LoginPage:

    _logo=".login_logo"

    _username_selector = "#user-name"
    _password_selector = "#password"
    _login_button_selector = "#login-button"

    def __init__(self, page: Page):
        self.page = page

    def login_page_loaded(self):
        expect(self.page.locator(self._logo)).to_be_visible()
        expect(self.page.locator(self._username_selector)).to_be_visible()
        expect(self.page.locator(self._password_selector)).to_be_visible()
        expect(self.page.locator(self._login_button_selector)).to_be_visible()

    def goto_login_page(self) -> None:
        self.page.goto("https://www.saucedemo.com/")

    def login(self, user: User) -> None:
        self.page.locator(self._username_selector).click()
        self.page.locator(self._username_selector).type(user.username)
        self.page.locator(self._password_selector).click()
        self.page.locator(self._password_selector).type(user.password)
        self.page.locator(self._login_button_selector).click()

File: wdio-cucumberjs/features/pageobjects/login.page.js
const Page = require('./page')

class LoginPage extends Page{

    get username(){
        return $('#user-name');
    }

    get password(){
        return $('#password');
    }

    get loginBtn(){
        return $('#login-button');
    }

    get errorMsg(){
        return $('h3[data-test="error"]');
    }

    async visitLoginPage() {
        await browser.url(`https://www.saucedemo.com`)
    }

    async confirmLoginPageDisplayed(){
        await expect(this.username).toBeExisting();
        await expect(this.password).toBeExisting();
        await expect(this.loginBtn).toBeExisting();
    }

    async authenticateUser(user, pass){
        if (user === undefined){
            user = 'standard_user';
        }
        if (pass === undefined){
            pass = 'secret_sauce';
        }

        await this.username.click();
        await this.username.setValue(user);
        await this.password.click();
        await this.password.setValue(pass);
        await this.loginBtn.click();
    }

    async validateInvalidLoginMsg(msg){
        expect(this.errorMsg).toHaveText(msg)
    }

}

module.exports = new LoginPage();